lib/swing-lib/src/main/java/org/triplea/ui/events/queue/ViewControllerEventQueue.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * <p>UIs publish UI events and Model publish model events. UI listen for model events and models
+ * listen for UI events. The model event callbacks (UI) are executed on the Swing EDT thread, and UI
+ * event callbacks are executed in a new thread (off of the EDT).
+ */
+@Setter
+public class ViewControllerEventQueue<
+    ViewDataT extends ViewData,
+    ViewEventTypeT,
+    ModelEventTypeT> {
+
+  private Collection<Object> controllers = new ArrayList<>();
+  private Collection<Object> views = new ArrayList<>();
+
+  public void addView(final Object view) {
+    views.add(view);
+  }
+
+  public void addController(final Object controller) {
+    controllers.add(controller);
+  }
+
+  /**
+   * When UI publishes an event, it should publish the full ui data state. This comes up as we will
+   * often attach action listeners that publish a ui event. If we attach an action listener to a
+   * text box then one would expect to emit the value of the text box and would _not_ expect to have
+   * to read the whole UI. Though, the UI shall publish full UI, otherwise the UI is assuming what
+   * data a controller needs, the decision of what to do next could easily depend on the value of
+   * other fields present in the UI.
+   */
+  public void publishUiEvent(final ViewEventTypeT uiEvent, final ViewDataT uiData) {
+    Preconditions.checkState(!controllers.isEmpty());
+
+    controllers.forEach(
+        view -> {
+          for (final Method method : view.getClass().getMethods()) {
+            if (method.getAnnotation(ViewListener.class) != null) {
+              try {
+                Preconditions.checkState(method.getParameterTypes().length == 2, method.getName());
+                Preconditions.checkState(method.getParameterTypes()[0].isAssignableFrom(uiEvent.getClass()));
+                Preconditions.checkState(method.getParameterTypes()[1].isAssignableFrom(uiEvent.getClass()));
+
+                method.invoke(view, uiEvent, uiData);
+
+              } catch (IllegalAccessException e) {
+                e.printStackTrace();
+              } catch (InvocationTargetException e) {
+                e.printStackTrace();
+              }
+              break;
+            }
+          }
+        });
+  }
+
+  /**
+   * Queue to pass events between a Swing UI and a view-model (AKA presentation layer).
+   *
+   * <p>UIs publish UI events and Model publish model events. UI listen for model events and models
+   * listen for UI events. The model event callbacks (UI) are executed on the Swing EDT thread, and
+   * UI event callbacks are executed in a new thread (off of the EDT).
+   */
+  public void publishModelEvent(
+      final ModelEventTypeT modelEvent, final UnaryOperator<ViewDataT> uiDataChange) {
+    Preconditions.checkState(!views.isEmpty());
+
+    views.forEach(
+        view -> {
+          for (final Method method : view.getClass().getMethods()) {
+            if (method.getAnnotation(ControllerListener.class) != null) {
```

Comment:
I think a compile-time alternative might be to define an interface that provides a method like this:
```java
List<BiConsumer<ModelEventTypeT, UnaryOperator<? extends ViewData>>> controllerListeners() {
  return List.of(
    this::method1,
    this::method2
  );
}
```

Not as elegant and with considerable overhead, but we get the compile-time checks.
Although I might be missing something